### PR TITLE
Make db-less possible, db-less test suite

### LINKED ISSUE #377: Refactor test suite to use GG Docker image
In our test suite, we use the official Kong Docker image and mount all required deps, plugins, etc into GG container.

Recently started some efforts to support official GG Docker image https://github.com/GluuFederation/https://github.com/GluuFederation/docker-gg

After some discussion with container team we decided next approach:
1. We will include docker-gg as git submodule in GG repo. It will allow as to build Docker image locally and don't depends on external autibuild workflow for our test suite.
2. The test suite should use locally build GG image within all tests.

Pros:
- simplify our test util a lot
- test hundreds of test scenarios on real image
- developers from containers team may use our test suite to test images with minimal efforts

### PR DESCRIPTION
**Pull request by altexy**
close #377

Almost all tests refactored to db-less deployment. Test running time is dramatically reduced.
One test with rate limiter plugin still uses Postgress.

Because of no possibility to change the Kong config during schema validation, the `method_path_tree` parameter of 3 plugins was refactored to use a new type of cache.

### CODE REVIEW

**Comment by altexy** [comment]
The test suite was refactored to use its own Docker image.
Docker image is built from sources upon every run.

**Comment by altexy** [comment]
> Test cases are failing please check the below log file

@ldeveloperl1985 can you, please test with, say, 10 seconds sleep, here https://github.com/GluuFederation/gluu-gateway/blob/9d5e24d2baef655e989464b61107edb6c7123174/t/lib/kong_utils.lua#L262

It looks timing issue.

File: kong/common/method-path-tree-cache.lua [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    return error("failed to create the cache: " .. (err or "unknown"))
+end
+
+return function(json_text)
+    local method_path_tree, stale_data = worker_cache:get(json_text)
+    if method_path_tree and not stale_data then
+        return method_path_tree
+    end
+
+    local oauth_scope_expression = cjson.decode(json_text)
```
[the remark this answers is not in the document]
Comment (altexy) [reply kong/common/method-path-tree-cache.lua]: sure

**Comment by ldeveloperl1985** [comment]
@altexy, Yes tested with 10 Sec and all test cases passed successfully.

**Comment by altexy** [comment]
Hi @ldeveloperl1985, I have added new test util - wait-for-http-ready.sh
It allows us to avoid sleep() to wait for HTTP server readiness.
Now all tests run less than 9 minutes on my laptop ;-)

Please, run tests on the latest branch code to ensure that I didn't forget anything.

**Comment by ldeveloperl1985** [comment]
@altexy Thank you, Test cases run successfully. 
One question, do we need these two files?
1. https://github.com/GluuFederation/gluu-gateway/blob/377_tests_docker_image/t/Dockerfile
2. https://github.com/GluuFederation/gluu-gateway/blob/377_tests_docker_image/t/run-flow.sh

**Comment by altexy** [comment]
> @altexy Thank you, Test cases run successfully.
> One question, do we need these two files?
> 
> 1. https://github.com/GluuFederation/gluu-gateway/blob/377_tests_docker_image/t/Dockerfile

Sure. It is Dockerfile of test runner.

> 2. https://github.com/GluuFederation/gluu-gateway/blob/377_tests_docker_image/t/run-flow.sh

I use it in my perf tests, not merged into master so far.

**Pull request merged**